dhall-docs/src/Dhall/Docs/CodeRenderer.hs

```diff
@@ -239,6 +239,28 @@ fragments = Data.List.sortBy sorter . removeUnusedDecls . Writer.execWriter . in
 
         Record (Map.toList -> l) -> handleRecordLike l
 
+        -- The parsed text of the import is located in it's `Note` constructor
+        Note src (Embed a) -> Writer.tell [SourceCodeFragment src $ ImportExpr a] >> return NoInfo
+
+        Note src (Var (V name index)) ->
+            case Context.lookup name index context of
+                Nothing -> return NoInfo
+                Just nameDecl@(NameDecl _ _ t) -> do
+                    Writer.tell [SourceCodeFragment src $ NameUse nameDecl]
+                    return t
+
+        Note src (BoolLit b) -> Writer.tell [SourceCodeFragment src $ StaticExpression _type] >> return NoInfo
+          where
+            _type = TypeFromSource $ case TypeCheck.typeOf (BoolLit b) of
+                Left _ -> fileAnIssue "typeOfStatic"
+                Right e -> e
+
+        Note src (NaturalLit n) -> Writer.tell [SourceCodeFragment src $ StaticExpression _type] >> return NoInfo
+          where
+            _type = TypeFromSource $ case TypeCheck.typeOf (NaturalLit n) of
+                Left _ -> fileAnIssue "typeOfStatic"
+                Right e -> e
```

Comment:
I recommend studying how `dhall-lsp-server` does this:

https://github.com/dhall-lang/dhall-haskell/blob/master/dhall-lsp-server/src/Dhall/LSP/Backend/Typing.hs

For example, one trick you can reuse from that code is to use the `subExpressions` `Traversal` to detect if a term has children or not when deciding whether or not to create a type tooltip